src/hotspot/share/code/dependencies.cpp

```diff
@@ -1342,6 +1389,175 @@ Klass* ConcreteMethodFinder::find_witness_anywhere(InstanceKlass* context_type)
+  // to enable dependency spot-checking and speed up the search.
+  LinkedConcreteMethodFinder(InstanceKlass* resolved_klass, Method* resolved_method, Method* uniqm = NULL) : AbstractClassHierarchyWalker(NULL) {
+    assert(UseVtableBasedCHA, "required");
+    assert(resolved_klass->is_linked(), "required");
+    assert(resolved_method->method_holder()->is_linked(), "required");
+    assert(!resolved_method->can_be_statically_bound(), "no vtable index available");
+
+    _resolved_klass  = resolved_klass;
+    _declaring_klass = resolved_method->method_holder();
+    _vtable_index    = compute_vtable_index(resolved_klass, resolved_method,
+                                            _do_itable_lookup); // out parameter
+    assert(_vtable_index >= 0, "invalid vtable index");
+
+    initialize(uniqm);
+  }
+
+  // Note:  If n==num_participants, returns NULL.
+  Method* found_method(uint n) {
+    assert(n <= num_participants(), "oob");
+    assert(participant(n) != NULL || n == num_participants(), "proper usage");
+    return _found_methods[n];
+  }
+};
+
+Klass* LinkedConcreteMethodFinder::find_witness_in(KlassDepChange& changes) {
+  Klass* type = changes.type();
+
+  assert(!is_participant(type), "only old classes are participants");
+
+  if (is_witness(type)) {
+    return type;
+  }
+  return NULL; // No witness found.  The dependency remains unbroken.
+}
+
+Klass* LinkedConcreteMethodFinder::find_witness_anywhere(InstanceKlass* context_type) {
+  for (CountingClassHierarchyIterator iter(context_type); !iter.done(); iter.next()) {
+    Klass* sub = iter.klass();
+    if (is_witness(sub)) {
+      return sub;
+    }
+    if (sub->is_instance_klass() && !InstanceKlass::cast(sub)->is_linked()) {
+      iter.skip_subclasses(); // ignore not yet linked classes
+    }
+  }
+  return NULL; // No witness found. The dependency remains unbroken.
+}
+
+bool LinkedConcreteMethodFinder::is_witness(Klass* k) {
+  if (is_participant(k)) {
+    return false; // do not report participant types
+  } else if (k->is_instance_klass()) {
+    InstanceKlass* ik = InstanceKlass::cast(k);
+    if (is_concrete_klass(ik)) {
+      Method* m = select_method(ik);
+      return record_witness(ik, m);
+    } else {
+      return false; // ignore non-concrete holder class
+    }
+  } else {
+    return false; // no methods to find in an array type
+  }
+}
+
+Method* LinkedConcreteMethodFinder::select_method(InstanceKlass* recv_klass) {
+  Method* selected_method = NULL;
+  if (_do_itable_lookup) {
+    assert(_declaring_klass->is_interface(), "sanity");
+    bool implements_interface; // initialized by method_at_itable_or_null()
+    selected_method = recv_klass->method_at_itable_or_null(_declaring_klass, _vtable_index,
+                                                           implements_interface); // out parameter
+    assert(implements_interface, "not implemented");
```

Comment:
Looking at `recv_klass->method_at_itable_or_null`, I wonder if there can be “holes” in the itable for missing methods.  They would lead to `AME` if called.  They might also trigger your assert here:  `assert(implements_interface, "not implemented")`.  Is there some reason that `select_method` cannot possibly encounter a missing method?

Answer to self:  I don't remember whether the JVM creates itable methods on the fly, but I suppose it does, so the code would see an synthetic abstract method.  (Decades ago we named those Miranda Methods because if you don't have a responding method "one will be provided for you".)  And itables are just aliases of vtable slices, so the miranda placed in the vtable will be seen also in the itable.

(Overall comment on this area of the code:  It looks great, much better than when I touched it last.  Thanks!)

Reply:
Regarding "holes" in itables: yes, it happens in practice. `select_method` is allowed to return a `NULL` and it is used by `LinkedConcreteMethodFinder` as a sentinel value for `AME` (when placed in `_found_methods` array). That's why `Dependencies::find_unique_concrete_method()` check `participant(0)`:
```
   Method* fm = wf.found_method(0);  // Will be NULL if num_parts == 0.
   Klass*   p = wf.participant(0);   // Will be NULL if num_parts == 0.
   ...
   if (Dependencies::is_concrete_method(m, ctxk)) {
     if (fm == NULL && p == NULL) {
       // It turns out that m was always the only implementation.
       fm = m;
     }
```

Now I think that it's not clear enough from the code. I'll elaborate on it in the additional comments.

Reply:
Regarding the assert: the itable should always be found. First, because only implementors are enumerated during the analysis (currently, subclasses of the unique direct implementor). Second, because itable stubs rely on itable contents when performing receiver subtype checks. So, all superinterfaces should have a corresponding itable present (even an empty one).